File: src/core/validators/Descriptor/DescriptorMapTypeItemsMustBeUnique.js
// @flow
import type SymbolTable from '../SymbolTable';
import { errorRuleBase } from '../ValidationRuleBase';
import { includeRuleBase } from '../ValidationRuleRepository';
import { MetaEdGrammar } from '../../../../src/grammar/gen/MetaEdGrammar';
import { findDuplicates } from '../ValidationHelper';

function getShortDescriptions(ruleContext: any) {
  return ruleContext.withMapType().enumerationItem().map(x => x.shortDescription().getText());
}

// eslint-disable-next-line no-unused-vars
function valid(ruleContext: any, symbolTable: SymbolTable): boolean {
  if (ruleContext.withMapType() == null) return true;
  const shortDescriptions = getShortDescriptions(ruleContext);
  if (shortDescriptions.length === 0) return true;
  return findDuplicates(shortDescriptions).length === 0;
}

// eslint-disable-next-line no-unused-vars
function failureMessage(ruleContext: any, symbolTable: SymbolTable): string {
  const identifier = ruleContext.descriptorName().getText();
  const duplicates = findDuplicates(getShortDescriptions(ruleContext));
  const joinString = '\', \'';
  return `Descriptor '${identifier}' declares duplicate item${duplicates.length > 1 ? 's' : ''} '${duplicates.join(joinString)}'.`;
}

const validationRule = errorRuleBase(valid, failureMessage);
// eslint-disable-next-line import/prefer-default-export
export const includeRule = includeRuleBase(MetaEdGrammar.RULE_descriptor, validationRule);

File: src/core/validators/Descriptor/DescriptorMapTypeItemsMustBeUnique.cs
using System.Linq;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.Descriptor
{
    public class DescriptorMapTypeItemsMustBeUnique : ValidationRuleBase<MetaEdGrammar.DescriptorContext>
    {
        private static string[] DuplicateShortDescriptions(MetaEdGrammar.DescriptorContext context)
        {
            // short descriptions are on the optional map type
            if (context.withMapType() == null) return new string[0];

            var shortDescriptions = context.withMapType().enumerationItem().Select(x => x.shortDescription().GetText());

            //group and filter duplicates
            return shortDescriptions.GroupBy(x => x).Where(group => group.Count() > 1).Select(group => group.Key).ToArray();
        }

        public override bool IsValid(MetaEdGrammar.DescriptorContext context)
        {
            return !DuplicateShortDescriptions(context).Any();
        }

        public override string GetFailureMessage(MetaEdGrammar.DescriptorContext context)
        {
            var identifier = context.descriptorName().GetText();
            var duplicateShortDescriptions = DuplicateShortDescriptions(context);

            return string.Format("Descriptor '{0}' declares duplicate item{2} '{1}'.", identifier, string.Join("', '", duplicateShortDescriptions), duplicateShortDescriptions.Count() > 1 ? "s" : string.Empty);
        }
    }
}
